// graphql-compose-elasticsearch/lib/resolvers/updateById.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

exports.default = createUpdateByIdResolver;
exports.getRecordITC = getRecordITC;

var _graphqlCompose = require('graphql-compose');

var _ElasticApiParser = require('../ElasticApiParser');

var _ElasticApiParser2 = _interopRequireDefault(_ElasticApiParser);

var _UpdateByIdOutput = require('../types/UpdateByIdOutput');

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function createUpdateByIdResolver(fieldMap, sourceTC, opts) {
  var _this = this;

  if (!fieldMap || !fieldMap._all) {
    throw new Error('First arg for Resolver updateById() should be fieldMap of FieldsMapByElasticType type.');
  }

  if (!sourceTC || sourceTC.constructor.name !== 'TypeComposer') {
    throw new Error('Second arg for Resolver updateById() should be instance of TypeComposer.');
  }

  var prefix = opts.prefix || 'Es';

  var parser = new _ElasticApiParser2.default({
    elasticClient: opts.elasticClient,
    prefix
  });

  var updateByIdFC = parser.generateFieldConfig('update', {
    index: opts.elasticIndex,
    type: opts.elasticType,
    _source: true
  });

  var argsConfigMap = {
    id: 'String!',
    record: getRecordITC(fieldMap).getTypeNonNull()
  };

  var type = (0, _UpdateByIdOutput.getUpdateByIdOutputTC)({ prefix, fieldMap, sourceTC });

  return new _graphqlCompose.Resolver({
    type,
    name: 'updateById',
    kind: 'mutation',
    args: argsConfigMap,
    resolve: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(rp) {
        var source, args, context, info, res, _ref2, _index, _type, _id, _version, result, get, _ref3, _source;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                source = rp.source, args = rp.args, context = rp.context, info = rp.info;


                args.body = {
                  doc: (0, _extends3.default)({}, args.record)
                };
                delete args.record;

                _context.next = 5;
                return updateByIdFC.resolve(source, args, context, info);

              case 5:
                res = _context.sent;
                _ref2 = res || {}, _index = _ref2._index, _type = _ref2._type, _id = _ref2._id, _version = _ref2._version, result = _ref2.result, get = _ref2.get;
                _ref3 = get || {}, _source = _ref3._source;
                return _context.abrupt('return', (0, _extends3.default)({
                  _id,
                  _index,
                  _type,
                  _version,
                  result
                }, _source));

              case 9:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, _this);
      }));

      function resolve(_x) {
        return _ref.apply(this, arguments);
      }

      return resolve;
    }()
  });
}

function getRecordITC(fieldMap) {
  var name = (0, _utils.getTypeName)('Record', {});
  var description = (0, _utils.desc)(`The record from Elastic Search`);
  return (0, _utils.getOrSetType)(name, function () {
    return _graphqlCompose.InputTypeComposer.create({
      name,
      description,
      fields: (0, _extends3.default)({}, fieldMap._all)
    });
  });
}
