// graphql-compose-elasticsearch/lib/elasticDSL/Query/TermLevel/Term.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getTermITC = getTermITC;

var _graphqlCompose = require('graphql-compose');

var _utils = require('../../../utils');

var _FieldNames = require('../../Commons/FieldNames');

function getTermITC() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var name = (0, _utils.getTypeName)('QueryTerm', opts);
  var description = (0, _utils.desc)(`
    Find documents which contain the exact term specified
    in the field specified.
    { fieldName: value } or { fieldName: { value: value, boost: 2.0 } }
    [Documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)
  `);

  var subName = (0, _utils.getTypeName)('QueryTermSettings', opts);
  var fields = (0, _FieldNames.getAllAsFieldConfigMap)(opts, (0, _utils.getOrSetType)(subName, function () {
    return _graphqlCompose.InputTypeComposer.create({
      name: subName,
      fields: {
        value: 'JSON!',
        boost: 'Float',
        fuzziness: 'Int',
        prefix_length: 'Int',
        max_expansions: 'Int'
      }
    });
  }));

  if (typeof fields === 'object') {
    return (0, _utils.getOrSetType)(name, function () {
      return _graphqlCompose.InputTypeComposer.create({
        name,
        description,
        fields
      });
    });
  }

  return {
    type: 'JSON',
    description
  };
}
