// graphql-compose-elasticsearch/mjs/elasticApiFieldConfig.mjs
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/* eslint-disable no-param-reassign */

import { GraphQLObjectType, GraphQLString } from 'graphql-compose/lib/graphql';
import elasticsearch from 'elasticsearch';
import ElasticApiParser from './ElasticApiParser';

const DEFAULT_ELASTIC_API_VERSION = '_default';

export function elasticApiFieldConfig(esClientOrOpts) {
  if (!esClientOrOpts || typeof esClientOrOpts !== 'object') {
    throw new Error('You should provide ElasticClient instance or ElasticClientConfig in first argument.');
  }

  if (isElasticClient(esClientOrOpts)) {
    return instanceElasticClient(esClientOrOpts);
  } else {
    return contextElasticClient(esClientOrOpts);
  }
}

function instanceElasticClient(elasticClient) {
  const apiVersion = elasticClient.transport._config.apiVersion || DEFAULT_ELASTIC_API_VERSION;
  const prefix = `ElasticAPI${apiVersion.replace('.', '')}`;

  const apiParser = new ElasticApiParser({
    elasticClient,
    prefix
  });

  return {
    description: `Elastic API v${apiVersion}`,
    type: new GraphQLObjectType({
      name: prefix,
      fields: apiParser.generateFieldMap()
    }),
    resolve: () => ({})
  };
}

function contextElasticClient(elasticConfig) {
  if (!elasticConfig.apiVersion) {
    elasticConfig.apiVersion = DEFAULT_ELASTIC_API_VERSION;
  }
  const { apiVersion } = elasticConfig;
  const prefix = `ElasticAPI${apiVersion.replace('.', '')}`;

  const apiParser = new ElasticApiParser({
    apiVersion,
    prefix
  });

  return {
    description: `Elastic API v${apiVersion}`,
    type: new GraphQLObjectType({
      name: prefix,
      fields: apiParser.generateFieldMap()
    }),
    args: {
      host: {
        type: GraphQLString,
        defaultValue: elasticConfig.host || 'http://user:pass@localhost:9200'
      }
    },
    resolve: (src, args, context) => {
      if (typeof context === 'object') {
        const opts = args.host ? _extends({}, elasticConfig, {
          host: args.host
        }) : elasticConfig;
        context.elasticClient = new elasticsearch.Client(opts);
      }
      return {};
    }
  };
}

function isElasticClient(obj) {
  if (obj instanceof elasticsearch.Client) {
    return true;
  }

  if (obj && obj.transport && obj.transport._config && obj.transport._config.__reused) {
    return true;
  }

  return false;
}
