// graphql-compose/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProjectionNode = exports.ProjectionType = exports.ResolveDebugOpts = exports.ResolverWrapTypeCb = exports.ResolverWrapArgsCb = exports.ResolverNextRpCb = exports.ResolverRpCb = exports.ResolverWrapCb = exports.ResolverOpts = exports.ResolverSortArgConfig = exports.ResolverSortArgFn = exports.ResolverFilterArgConfig = exports.ResolverFilterArgFn = exports.ResolverKinds = exports.ResolveParams = exports.ComposeInputObjectTypeConfig = exports.ComposeInputFieldConfigMap = exports.ComposeInputFieldConfigAsObject = exports.ComposeInputFieldConfig = exports.ComposeInputType = exports.filterByDotPaths = exports.deepmerge = exports.toDottedObject = exports.convertInputObjectFieldOpts = exports.toInputObjectType = exports.GraphQLJSON = exports.GraphQLGeneric = exports.GraphQLBuffer = exports.GraphQLDate = exports.getFlatProjectionFromAST = exports.getProjectionFromAST = exports.TypeStorage = exports._ProtoTypeMapper = exports._ProtoResolver = exports._ProtoEnumTypeComposer = exports._ProtoInputTypeComposer = exports._ProtoTypeComposer = exports.TypeMapper = exports.Resolver = exports.EnumTypeComposer = exports.InputTypeComposer = exports.TypeComposer = exports.SchemaComposer = exports.schemaComposer = exports.GQC = exports.graphql = undefined;

var _TypeComposer = require('./TypeComposer');

Object.defineProperty(exports, '_ProtoTypeComposer', {
  enumerable: true,
  get: function get() {
    return _TypeComposer.TypeComposer;
  }
});

var _InputTypeComposer = require('./InputTypeComposer');

Object.defineProperty(exports, '_ProtoInputTypeComposer', {
  enumerable: true,
  get: function get() {
    return _InputTypeComposer.InputTypeComposer;
  }
});

var _EnumTypeComposer = require('./EnumTypeComposer');

Object.defineProperty(exports, '_ProtoEnumTypeComposer', {
  enumerable: true,
  get: function get() {
    return _EnumTypeComposer.EnumTypeComposer;
  }
});

var _Resolver = require('./Resolver');

Object.defineProperty(exports, '_ProtoResolver', {
  enumerable: true,
  get: function get() {
    return _Resolver.Resolver;
  }
});

var _TypeMapper = require('./TypeMapper');

Object.defineProperty(exports, '_ProtoTypeMapper', {
  enumerable: true,
  get: function get() {
    return _TypeMapper.TypeMapper;
  }
});

var _TypeStorage = require('./TypeStorage');

Object.defineProperty(exports, 'TypeStorage', {
  enumerable: true,
  get: function get() {
    return _TypeStorage.TypeStorage;
  }
});

var _projection = require('./utils/projection');

Object.defineProperty(exports, 'getProjectionFromAST', {
  enumerable: true,
  get: function get() {
    return _projection.getProjectionFromAST;
  }
});
Object.defineProperty(exports, 'getFlatProjectionFromAST', {
  enumerable: true,
  get: function get() {
    return _projection.getFlatProjectionFromAST;
  }
});

var _type = require('./type');

Object.defineProperty(exports, 'GraphQLDate', {
  enumerable: true,
  get: function get() {
    return _type.GraphQLDate;
  }
});
Object.defineProperty(exports, 'GraphQLBuffer', {
  enumerable: true,
  get: function get() {
    return _type.GraphQLBuffer;
  }
});
Object.defineProperty(exports, 'GraphQLGeneric', {
  enumerable: true,
  get: function get() {
    return _type.GraphQLGeneric;
  }
});
Object.defineProperty(exports, 'GraphQLJSON', {
  enumerable: true,
  get: function get() {
    return _type.GraphQLJSON;
  }
});

var _toInputObjectType = require('./utils/toInputObjectType');

Object.defineProperty(exports, 'toInputObjectType', {
  enumerable: true,
  get: function get() {
    return _toInputObjectType.toInputObjectType;
  }
});
Object.defineProperty(exports, 'convertInputObjectFieldOpts', {
  enumerable: true,
  get: function get() {
    return _toInputObjectType.convertInputObjectFieldOpts;
  }
});

var _misc = require('./utils/misc');

Object.keys(_misc).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _misc[key];
    }
  });
});

var _is = require('./utils/is');

Object.keys(_is).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _is[key];
    }
  });
});

var _graphqlVersion = require('./utils/graphqlVersion');

Object.keys(_graphqlVersion).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _graphqlVersion[key];
    }
  });
});

var _toDottedObject = require('./utils/toDottedObject');

Object.defineProperty(exports, 'toDottedObject', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_toDottedObject).default;
  }
});

var _deepmerge = require('./utils/deepmerge');

Object.defineProperty(exports, 'deepmerge', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_deepmerge).default;
  }
});

var _filterByDotPaths = require('./utils/filterByDotPaths');

Object.defineProperty(exports, 'filterByDotPaths', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_filterByDotPaths).default;
  }
});
Object.defineProperty(exports, 'ComposeInputType', {
  enumerable: true,
  get: function get() {
    return _InputTypeComposer.ComposeInputType;
  }
});
Object.defineProperty(exports, 'ComposeInputFieldConfig', {
  enumerable: true,
  get: function get() {
    return _InputTypeComposer.ComposeInputFieldConfig;
  }
});
Object.defineProperty(exports, 'ComposeInputFieldConfigAsObject', {
  enumerable: true,
  get: function get() {
    return _InputTypeComposer.ComposeInputFieldConfigAsObject;
  }
});
Object.defineProperty(exports, 'ComposeInputFieldConfigMap', {
  enumerable: true,
  get: function get() {
    return _InputTypeComposer.ComposeInputFieldConfigMap;
  }
});
Object.defineProperty(exports, 'ComposeInputObjectTypeConfig', {
  enumerable: true,
  get: function get() {
    return _InputTypeComposer.ComposeInputObjectTypeConfig;
  }
});
Object.defineProperty(exports, 'ResolveParams', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolveParams;
  }
});
Object.defineProperty(exports, 'ResolverKinds', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverKinds;
  }
});
Object.defineProperty(exports, 'ResolverFilterArgFn', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverFilterArgFn;
  }
});
Object.defineProperty(exports, 'ResolverFilterArgConfig', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverFilterArgConfig;
  }
});
Object.defineProperty(exports, 'ResolverSortArgFn', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverSortArgFn;
  }
});
Object.defineProperty(exports, 'ResolverSortArgConfig', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverSortArgConfig;
  }
});
Object.defineProperty(exports, 'ResolverOpts', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverOpts;
  }
});
Object.defineProperty(exports, 'ResolverWrapCb', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverWrapCb;
  }
});
Object.defineProperty(exports, 'ResolverRpCb', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverRpCb;
  }
});
Object.defineProperty(exports, 'ResolverNextRpCb', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverNextRpCb;
  }
});
Object.defineProperty(exports, 'ResolverWrapArgsCb', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverWrapArgsCb;
  }
});
Object.defineProperty(exports, 'ResolverWrapTypeCb', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolverWrapTypeCb;
  }
});
Object.defineProperty(exports, 'ResolveDebugOpts', {
  enumerable: true,
  get: function get() {
    return _Resolver.ResolveDebugOpts;
  }
});
Object.defineProperty(exports, 'ProjectionType', {
  enumerable: true,
  get: function get() {
    return _projection.ProjectionType;
  }
});
Object.defineProperty(exports, 'ProjectionNode', {
  enumerable: true,
  get: function get() {
    return _projection.ProjectionNode;
  }
});

var _graphql = require('./graphql');

var graphql = _interopRequireWildcard(_graphql);

var _SchemaComposer = require('./SchemaComposer');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*  strict */

exports.graphql = graphql;


const schemaComposer = new _SchemaComposer.SchemaComposer();
const GQC = schemaComposer;
const TypeComposer = GQC.TypeComposer;
const InputTypeComposer = GQC.InputTypeComposer;
const EnumTypeComposer = GQC.EnumTypeComposer;
const Resolver = GQC.Resolver;
const TypeMapper = GQC.typeMapper;
exports.GQC = GQC;
exports.schemaComposer = schemaComposer;
exports.SchemaComposer = _SchemaComposer.SchemaComposer;
exports.TypeComposer = TypeComposer;
exports.InputTypeComposer = InputTypeComposer;
exports.EnumTypeComposer = EnumTypeComposer;
exports.Resolver = Resolver;
exports.TypeMapper = TypeMapper;

// Export Composers' prototype classes for instance check in plugins
