// graphql-compose-elasticsearch/lib/types/SearchOutput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSearchOutputTC = getSearchOutputTC;

var _graphqlCompose = require('graphql-compose');

var _utils = require('../utils');

var _Shards = require('./Shards');

var _Shards2 = _interopRequireDefault(_Shards);

var _SearchHitItem = require('./SearchHitItem');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getSearchOutputTC() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var name = (0, _utils.getTypeName)('SearchOutput', opts);
  var nameHits = (0, _utils.getTypeName)('SearchHits', opts);

  return (0, _utils.getOrSetType)(name, function () {
    return _graphqlCompose.TypeComposer.create({
      name,
      fields: {
        took: 'Int',
        timed_out: 'Boolean',
        _shards: (0, _Shards2.default)(opts),
        hits: (0, _utils.getOrSetType)(nameHits, function () {
          return _graphqlCompose.TypeComposer.create({
            name: nameHits,
            fields: {
              total: 'Int',
              max_score: 'Float',
              hits: [(0, _SearchHitItem.getSearchHitItemTC)(opts)]
            }
          });
        }),
        aggregations: 'JSON'
      }
    });
  });
}
